### PR TITLE
feat(caching): serve stale values on factory failure

### PR DESCRIPTION
**Pull request by xshaheen**
Closes #373

## Summary

Factory-backed cache reads can now survive a transient downstream failure by serving the last-known-good value while the entry is still physically retained. The behavior is per-entry opt-in through `CacheEntryOptions`, marks the activating response with `CacheValue<T>.IsStale`, and throttles repeated factory retries without extending the physical stale window.

The resilience flow is centralized in `Headless.Caching.Core` so Memory, Redis, and Hybrid share the same state machine. Normal reads remain logical-governed: once logical expiration passes, `GetAsync`/bulk reads/`ExistsAsync` miss, while `GetOrAddAsync` can still consume the physical reserve for fail-safe. Redis keeps the key TTL aligned with physical expiration, and Hybrid now checks L2 before falling back to a stale L1 reserve.

| Area | Behavior |
| --- | --- |
| Abstractions | `CacheEntryOptions` adds fail-safe duration/throttle knobs; `CacheValue<T>` exposes `IsStale`. |
| Core | `FactoryCacheCoordinator` owns locking, double-check, factory execution, stale activation, and best-effort restamp. |
| Memory | Stores logical and physical expiration separately; value reads use logical expiration while eviction/key presence use physical. |
| Redis | Frame metadata drives logical reads; key TTL follows physical retention for fail-safe entries. |
| Hybrid | Composes L1/L2 store metadata, caps L1 retention by local expiration, and prefers fresh L2 over stale L1. |

Docs were updated for `docs/llms/caching.md`, package READMEs, and the completed implementation plan.

## Review Hardening

Following a multi-agent code review, the coordinator and providers were hardened against several edge cases (all covered by the new tests below):

- **Fresh-value integrity** — only the factory call sits inside the fail-safe `catch`; a store-write failure on the fresh path now propagates instead of silently discarding the just-computed value and returning stale.
- **Throttle correctness** — a fail-safe stale candidate must carry a physical expiration, so the throttle restamp always lands (no factory hammering via null-physical entries).
- **Hybrid L1 promotion** — only logically-fresh L2 entries are promoted into L1 on reads (stops stale-read write amplification and overwriting a newer L1 reserve); fail-safe activation still refreshes L1 with the throttled entry, matching FusionCache parity and KTD-4. Legacy null-timestamp L2 entries are promoted bounded by the local ceiling.
- **Cancellation semantics** — activation is gated on caller-token identity (`OperationCanceledException.CancellationToken`), so an unrelated linked/internal timeout still activates fail-safe while genuine caller cancellation always propagates (KTD-7). The throttle restamp is best-effort (`CancellationToken.None`, unconditional swallow) so the stale value is always returned once activation is decided.
- **Read robustness** — `InMemoryCache.TryGetEntryAsync` mirrors the `GetAsync` deserialization guard (a corrupt entry no longer drops the stale candidate) and uses `PeekValue` to avoid clone/LRU side effects on the coordinator's double-check read.
- **Contract** — `IFactoryCacheStore.SetEntryAsync<T>` takes `T?` (honest nullability), engine types carry `[PublicAPI]`, `CacheValue<T>` enforces `IsStale ⇒ HasValue`, and the expiry boundary (`<= now`) is consistent across Core, Memory, Redis, and the eviction loop.

## Tests

- `make test-project TEST_PROJECT=tests/Headless.Caching.Abstractions.Tests.Unit`
- `make test-project TEST_PROJECT=tests/Headless.Caching.Core.Tests.Unit`
- `make test-project TEST_PROJECT=tests/Headless.Caching.Memory.Tests.Unit` (143 passed)
- `make test-project TEST_PROJECT=tests/Headless.Caching.Hybrid.Tests.Unit` (31 passed — incl. new `HybridCacheFailSafeTests`: L1/L2 stale serve, L2-down fallback, read-path promotion guard, throttle-refreshes-L1)
- `make test-project TEST_PROJECT=tests/Headless.Caching.Redis.Tests.Integration` (129 passed — incl. new `RedisCacheFailSafeTests`: R1/R4 stale serve, TTL=physical, logical-governed `GetAsync`/`ExistsAsync`/`GetExpirationAsync`, beyond-physical propagation)
- `make build` (0 warnings, 0 errors)

> Note: `RedisCacheFailSafeTests` adds the first provider-level coverage of R4 (store-unavailable → fail-safe). Redis integration tests require Docker.

## Post-Deploy Monitoring & Validation

Validation window: first 24 hours after package adoption in a service that enables fail-safe.
Owner: adopting service owner.

Healthy signals:
- Warning logs with event name `CacheFailSafeActivated` are rare and correlate with known downstream failures.
- Error rates at endpoints backed by fail-safe cache entries do not spike during downstream incidents.
- Redis memory and key TTL distribution stay within expected bounds for the configured `FailSafeMaxDuration`.

Failure signals and mitigation:
- Sustained `CacheFailSafeActivated` logs for one key or prefix indicate a hard-down dependency; reduce `FailSafeMaxDuration`, disable fail-safe for that entry, or fix the factory dependency.
- Redis memory growth beyond expected cache reserve size indicates physical retention is too long for the workload; lower `FailSafeMaxDuration` or add key-specific bounds.
- Unexpected stale responses during the throttle window should be mitigated by lowering `FailSafeThrottleDuration` or disabling fail-safe for the affected key until the factory is healthy.

Suggested log searches:
- `CacheFailSafeActivated`
- `CacheFailSafeRestampFailed`
- `CacheStoreReadFailed`
- `FailedToReadFromL2Cache`

### CODE REVIEW

File: src/Headless.Caching.Core/FactoryCacheCoordinator.cs [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+        {
+            return CacheValue<T>.NoValue;
+        }
+
+        return entry.IsNull
+            ? new CacheValue<T>(default, hasValue: true, isStale)
+            : new CacheValue<T>(entry.Value, hasValue: true, isStale);
+    }
+
+    private static bool _IsFresh<T>(CacheStoreEntry<T> entry, DateTime now) => entry.IsFresh(now);
```
Comment (xshaheen) [inline src/Headless.Caching.Core/FactoryCacheCoordinator.cs]: why it needed?

Comment (xshaheen) [reply src/Headless.Caching.Core/FactoryCacheCoordinator.cs]: > why it needed?

Not needed anymore. Originally `_IsFresh` held freshness logic inline (physical-presence + logical-expiration checks), but that was later moved into `CacheStoreEntry<T>.IsFresh(now)`, leaving the wrapper as a trivial passthrough. Removed — both call sites now call `entry.IsFresh(now)` directly.
